### ISSUE TITLE
In files.nix, allow `executable` on a directory to make all the leaves executable

### ISSUE THREAD

**Issue by sellout** [nix-community/home-manager#3594]
I was hoping to have something like this do what I want:
```nix
{
    home.file.".local/bin" = {
      executable = true;
      source = ../home/.local/bin;
    };
}
```

However, when the source is a directory, `executable` has no effect (and I couldn’t find documentation on that behavior). I was hoping it would symlink the directory, and all the files in the store would be `+x`.

**Comment by ncfavier** [comment]
I think that would be surprising. "Executable directory" isn't really a thing (in fact the executable bit has a completely different meaning on directories).

Is there any reason you can't make the files in `../home/.local/bin` executable?

You could also have an entry per file (use `builtins.readDir ../home/.local/bin` to generate the entries automatically). Then `executable` would work.

**Comment by sellout** [comment]
I agree it’s a bit magical, I considered not opening this issue, because I’m also on the fence. And I guess it technically follows the mode settings as is, making the directory “executable” or not, but man, file modes are such a wart, especially applied to directories.

Basically, I was trying to achieve a couple things:
* I like to not have the executable bits set on binaries in the repo, because I want to make sure they’re “installed” properly when they’re being used, and not being run from the repo instead of the correct location;
* being able to simply add a new script to the right place in the repo and know that it will be installed, rather than having to repeat myself and add noise to the config; and
* having ~/.local/bin be not-writable, so binaries can’t be added directly by accident, rather than managed by home-manager.

So a mix of system consistency and reducing friction.

More broadly, I like that home-manager makes my system more orderly – I like using it to get in the way of my laziness, where I copy files around, forget where they are or what machine they’re on, and generally lose track of how things are organized.

I suppose this just means adding a new function, like `directory-of-executables = target: source: …` and then
```nix
{
  home.file = {
    other files ...
  } // directory-of-executables ".local/bin" ../home/.local/bin;
}
```
which is a minor annoyance at worst.
